base_model_restrict_update/models/ir_model_access.py

```diff
@@ -0,0 +1,35 @@
+# Copyright 2021 Quartile Limited
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from odoo import api, models, tools
+from odoo.exceptions import AccessError
+from odoo.tools.translate import _
+
+
+class IrModelAccess(models.Model):
+    _inherit = "ir.model.access"
+
+    @api.model
+    @tools.ormcache_context(
+        "self._uid", "model", "mode", "raise_exception", keys=("lang",)
+    )
+    def check(self, model, mode="read", raise_exception=True):
+        if self.env.su:
+            return True
+        res = super().check(model, mode, raise_exception)
+        self._cr.execute(
+            "SELECT restrict_update FROM ir_model WHERE model = %s", (model,)
+        )
+        query_res = self._cr.dictfetchone()
+        if (
+            query_res["restrict_update"]
+            and mode != "read"
+            and not self.env.user.unrestrict_model_update
+        ):
+            if raise_exception:
+                raise AccessError(
+                    _("You are only allowed to read this record. (%s - %s)")
+                    % (model, mode)
+                )
+            return False
```

Comment:
```suggestion
        if mode == "read":
            return res
        self._cr.execute(
            "SELECT is_readonly_user FROM res_users WHERE id = %s", (selv.env.uid,)
        )
        user_query_res = self._cr.dictfetchone()
        if user_query_res["is_readonly_user"]:
            if raise_exception:
                self._raise_access_error()
            return False
        self._cr.execute(
            "SELECT restrict_update FROM ir_model WHERE model = %s", (model,)
        )
        model_query_res = self._cr.dictfetchone()
        if (
            model_query_res["restrict_update"]
            and not self.env.user.unrestrict_model_update
        ):
            if raise_exception:
                self._raise_access_error()
            return False
```

@kittiu What I had in mind is something like this, given that `is_readonly_user` is newly added to res.users.  Does this make sense?

Reply:
IC what you mean, is this correct?

* If user is marked as "readonly", regardless of any model restriction or user permit, just don't allow update.

In this case, there is nothing to do with model restriction, right? (what I propose was to still use the model restriction ways, but yes, it could be too restricted).

I would think, we should have another module that do only this, i..e, `base_user_readonly`. This will be simple module, which of course can work along side with base_module_restrict_update`.

What do you think?

Reply:
You understand me correctly, Kitti.

> I would think, we should have another module that do only this, i..e, `base_user_readonly`. This will be simple module, which of course can work along side with `base_module_restrict_update`.
> 
> What do you think?

While I think it makes sense to create a separate module `base_user_readonly`, I can also think of a couple of benefits to keep both functions in one module:

1. It makes it easier to avoid test conflicts.
2. Restrict/unrestrict toggle button can be hidden when user is a readonly user.

That said I do not have strong preferences.

Reply:
OK Yoshi. You are the original author, I would go by your will :).
For me, I am ok both way.
I will create another PR, and let you know when it is ready.